### PR TITLE
Cs 182 convert validated registration to apprentice

### PR DESCRIPTION
**Pull request by Pauljgraham**

### CODE REVIEW

File: src/SFA.DAS.ApprenticeCommitments/Application/Commands/VerifyRegistrationCommand/VerifyRegistrationCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+using SFA.DAS.ApprenticeCommitments.Exceptions;
+using SFA.DAS.ApprenticeCommitments.Models;
+
+namespace SFA.DAS.ApprenticeCommitments.Application.Commands.VerifyRegistrationCommand
+{
+    public class VerifyRegistrationCommandHandler : IRequestHandler<VerifyRegistrationCommand>
+    {
+        private readonly IRegistrationRepository _registrationRepository;
+        private readonly IApprenticeRepository _apprenticeRepository;
+        private readonly IApprenticeshipRepository _apprenticeshipRepository;
```
[the remark this answers is not in the document]
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Application/Commands/VerifyRegistrationCommand/VerifyRegistrationCommandHandler.cs:15]: As discussed we will use this as is and look to refactor in the next story

File: src/SFA.DAS.ApprenticeCommitments/Models/ApprenticeModel.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+namespace SFA.DAS.ApprenticeCommitments.Models
+{
+    public class ApprenticeModel
+    {
+        public long? Id { get; internal set; }
+        public string FirstName { get; set; }
+        public string LastName { get; set; }
+        public Guid UserIdentityId { get; set; }
+        public string Email { get; set; }
+        public DateTime DateOfBirth { get; set; }
```
[the remark this answers is not in the document]
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Models/ApprenticeModel.cs:13]: As discussed we will use this as is and look to refactor in the next story

File: src/SFA.DAS.ApprenticeCommitments/Models/RegistrationModel.cs [hunk 3]
```diff
@@ -11,5 +8,9 @@ public class RegistrationModel
         public long ApprenticeshipId { get; set; }
         public string Email { get; set; }
         public DateTime? CreatedOn { get; private set; }
+        public Guid? UserIdentityId { get; set; }
+        public long? ApprenticeId { get; set; }
```
[the remark this answers is not in the document]
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Models/RegistrationModel.cs:12]: As discussed we will use this as is and look to refactor in the next story

File: src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/VerifyRegistrationSteps.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            var content = await _context.Api.Response.Content.ReadAsStringAsync();
+            var errors = JsonConvert.DeserializeObject<List<ErrorItem>>(content);
+
+            errors.Count(x => x.PropertyName == "RegistrationId").Should().Be(1);
+            errors.Count(x => x.PropertyName == "UserIdentityId").Should().Be(1);
+            errors.Count(x => x.PropertyName == "FirstName").Should().Be(1);
+            errors.Count(x => x.PropertyName == "LastName").Should().Be(1);
+            errors.Count(x => x.PropertyName == "DateOfBirth").Should().Be(1);
+            errors.Count(x => x.PropertyName == "Email").Should().Be(1);
+        }
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/VerifyRegistrationSteps.cs:174]: I like that, will add it

File: src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/VerifyRegistrationSteps.cs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+        [Then(@"response contains the not found error message")]
+        public async Task ThenResponseContainsTheNotFoundErrorMessage()
+        {
+            var content = await _context.Api.Response.Content.ReadAsStringAsync();
+            var errors = JsonConvert.DeserializeObject<List<ErrorItem>>(content);
+
+            errors.Count().Should().Be(1);
+            errors[0].ErrorMessage.Should().Contain("not found");
+            errors[0].ErrorMessage.Should().Contain("Registration");
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/VerifyRegistrationSteps.cs:193]: Well this is a domain error the format of the error message is "Registration 9999-9999...99 not found"

and I need to identify we raised the correct  error. Perhaps we should be using Error numbers to allow us to identify explicit types of errors

File: src/SFA.DAS.ApprenticeCommitments.Api/Extensions/ApplicationBuilderExtensions.cs [hunk 6]
```diff
@@ -27,6 +28,11 @@ static async Task Handler(HttpContext context)
                         context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
                         await context.Response.WriteAsync(CreateErrorResponse(ex.Errors));
                     }
+                    else if (contextFeature.Error is DomainException domainException)
+                    {
+                        context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+                        await context.Response.WriteAsync(CreateErrorResponse(domainException.Message));
+                    }
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Api/Extensions/ApplicationBuilderExtensions.cs:35]: Yes, I don't think we would have more than these two types of exceptions being passed back. Did consider a switch statement

File: src/SFA.DAS.ApprenticeCommitments.Database/Tables/Apprentice.sql [hunk 7]
```diff
@@ -0,0 +1,16 @@
+﻿CREATE TABLE [dbo].[Apprentice]
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Database/Tables/Apprentice.sql:1]: Personally I prefer not to use plurals on table tables. But if you have a strong view on it then I'm happy to change to plurals, but we have to be consistent

File: src/SFA.DAS.ApprenticeCommitments.Database/Tables/Apprenticeship.sql [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿CREATE TABLE [dbo].[Apprenticeship]
+(
+	[Id] BIGINT IDENTITY(10000,1) NOT NULL PRIMARY KEY,
+	[ApprenticeId] BIGINT NOT NULL,
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Database/Tables/Apprenticeship.sql:4]: Good spot

File: src/SFA.DAS.ApprenticeCommitments/Data/Models/Apprentice.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace SFA.DAS.ApprenticeCommitments.Data.Models
+{
+    [Table("Apprentice")]
+    public class Apprentice
+    {
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments/Data/Models/Apprentice.cs:9]: I like adding it to the model as suggested, however I would like to explicitly keep the ValueGeneratedOnAdd() as it should expected intention and considering we are using SQLite in Accept Tests, I worry about inferring behaviour

File: src/SFA.DAS.ApprenticeCommitments/Data/RegistrationRepository.cs [hunk 10]
```diff
@@ -34,5 +34,19 @@ public async Task<RegistrationModel> Get(Guid registrationId)
...
+            var db = _dbContext.Value;
+            var entity = await db.Registrations.FirstOrDefaultAsync(x => x.Id == registrationId);
+
+            entity.ApprenticeId = apprenticeId;
+            entity.UserIdentityId = userIdentityId;
+
+            db.Attach(entity);
+            db.Entry(entity).Property(x => x.ApprenticeId).IsModified = true;
+            db.Entry(entity).Property(x => x.UserIdentityId).IsModified = true;
+            db.Update(entity);
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments/Data/RegistrationRepository.cs:49]: Well i ended up putting it in after having issues with testing, but the article I read implied that if we don't do this all fields are updated

File: src/SFA.DAS.ApprenticeCommitments/Models/ApprenticeshipModel.cs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿namespace SFA.DAS.ApprenticeCommitments.Models
+{
+    public class ApprenticeshipModel
+    {
+        public long? Id { get; set; }
+        public long ApprenticeId { get; set; }
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments/Models/ApprenticeshipModel.cs:6]: Agree, but can we do this change when we rework the data layer

File: src/SFA.DAS.ApprenticeCommitments.Database/Tables/Apprenticeship.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Database/Tables/Apprenticeship.sql:4]: done

File: src/SFA.DAS.ApprenticeCommitments/Data/RegistrationRepository.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments/Data/RegistrationRepository.cs:49]: Will try it out

[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments/Data/RegistrationRepository.cs:49]: yes the article is correct if we don't do this all fields are updated

**Pull request merged**